runtime/gc_glue_java/FlattenedArrayObjectScanner.hpp

```diff
@@ -0,0 +1,255 @@
+/*******************************************************************************
+ * Copyright (c) 2020, 2020 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#if !defined(FLATTENEDARRAYOBJECTSCANNER_HPP_)
+#define FLATTENEDARRAYOBJECTSCANNER_HPP_
+
+#include "j9.h"
+#include "j9cfg.h"
+#include "modron.h"
+
+#include "objectdescription.h"
+#include "ArrayObjectModel.hpp"
+#include "GCExtensionsBase.hpp"
+#include "HeadlessMixedObjectScanner.hpp"
+#include "IndexableObjectScanner.hpp"
+
+class GC_FlattenedArrayObjectScanner : public GC_HeadlessMixedObjectScanner
+{
+	/* Data Members */
+private:
+
+	MM_EnvironmentBase *_env;
```

Comment:
I'd rather have this (and all parent) classes cache and pass through init method GCExtentions then Environment, since they don't seem  to use anything that is thread specific. This sort of makes assumption that thread that created the scanner will not be passed to another thread for usage (which might not be true - it might be passed through global scan queue).

@dmitripivkine your opinion?

Anyhow, that's a separate issue (would require OMR change, as well).

Reply:
Yes, I agree. I noticed this and it might be a source of the problem potentially

Reply:
Created an issue for this: https://github.com/eclipse/omr/issues/5018